docs/new-feature-tutorial.rst

```diff
@@ -2,84 +2,178 @@
+The blackbox tests are run in a headless browser using Casper.js and are located
+in ``zerver/tests/frontend/tests/``. The unit tests use Node's ``assert`` module
+are located in ``zerver/tests/frontend/node/``. For more information on testing
+see the :doc:`testing documentation <testing>`.
 
-You need to change various things on the front end.  In this case, the relevant files
-are `static/js/server_events.js`, `static/js/admin.js`, `static/styles/zulip.css
-and `static/templates/admin_tab.handlebars`.
+Example Feature
+===============
 
+This example describes the process of adding a new setting to Zulip:
+a flag that restricts inviting new users to admins only (the default behavior
+is that any user can invite other users).
+
+First, update the database and model to store the new setting. Add a 
+new boolean field, ``realm_invite_by_admins_only``, to the Realm model in
+``zerver/models.py``.
+
+Then create a Django migration that adds a new field, ``invite_by_admins_only``,
+to the ``zerver_realm`` table.
+
+In ``zerver/lib/actions.py``, create a new function named 
+``do_set_realm_invite_by_admins_only``. This function will update the database
+and trigger an event to notify clients when this setting changes. ::
+
+  # zerver/lib/actions.py
+
+  def do_set_realm_invite_by_admins_only(realm, invite_by_admins_only):
+    realm.invite_by_admins_only = invite_by_admins_only
+    realm.save(update_fields=['invite_by_admins_only'])
+    event = dict(
+      type="realm",
+      op="update",
+      property='invite_by_admins_only',
+      value=invite_by_admins_only,
+    )
+    send_event(event, active_user_ids(realm))
+    return {}
+
+You then need to add code that will handle the event and update the application state. In ``zerver/lib/actions.py`` update the ``fetch_initial_state`` and ``apply_events`` functions. ::
+
+  def fetch_initial_state_data(user_profile, event_types, queue_id):
+    # ...
+    state['realm_invite_by_admins_only'] = user_profile.realm.invite_by_admins_only`
+
+In this case you don't need to change ``apply_events`` because there is already code that will correctly handle this event: ::
+
+  def apply_events(state, events, user_profile):
+    for event in events:
+      # ...
+      elif event['type'] == 'realm':
+         field = 'realm_' + event['property']
+         state[field] = event['value']
+
+This feature adds a new parameter that should be sent to clients when the
+application loads and be accessible via JavaScript, so add it the initial state
+in ``zerver/views/__init__.py``. ::
+
+  # zerver/views/__init__.py
+
+  def home(request):
+    # ...
+    page_params = dict(
+      # ...
+      realm_invite_by_admins_only = register_ret['realm_invite_by_admins_only'],
+      # ...
+    )
+
+Since this feature also adds a checkbox to the admin page, and adds a new property the Realm model that can be modified from there, you also need to make changes to the ``update_realm`` function in the same file: ::
+
```

Comment:
We should make clear that structurally what we need to do here is make a view for the client to call the zerver/lib/actions.py function to change the state, and that there's already a convenient update_realm view that does this for several related flags, so we're just going to extend that function.